### ISSUE TITLE
Same query SELECT COUNT ... WHERE NOT IN returns different values on repeated execution

### ISSUE THREAD

**Issue by alexeykobzev** [ClickHouse/ClickHouse#36847]
> You have to provide the following information whenever possible.

I've two tables names `session` and `slice` with ReplacingMergeTree Engine

```
CREATE TABLE slice
(
    `uid` Int64,
    `start` DateTime64(3),
    `stop` DateTime64(3),
    `ip` String,
    `ip_operator` Int32,
    `asn` Int32,
    `ip_access_type` Int8,
    `ip_index` Int32,
    `conn_index` Int32,
    `cell_index` Int32,
    `geocell` Int64,
    `geocell_index` Int32,
    `host` String,
    `cdn_index` Int32,
    `quality` Int16,
    `quality_index` Int32,
    `play_time` Int32,
)
ENGINE = ReplacingMergeTree
PARTITION BY toYYYYMM(start)
ORDER BY (start, uid, ip_index, conn_index, cell_index, geocell_index, cdn_index, quality_index)
SETTINGS index_granularity = 8192
```

```
CREATE TABLE session
(
    `start` DateTime64(3),
    `stop` DateTime64(3),
    `service` Int32,
    `cid` Array(Int8),
    `sid` Array(Int8),
    `uid` Int64,
    `version` Int16,
    `device_os` String,
    `device_model` String,
    `device_build` String,
    `content_duration` Int32,
    `content_id` String,
    `term_cause` Int8
)
ENGINE = ReplacingMergeTree
PARTITION BY toYYYYMM(start)
ORDER BY (start, uid)
SETTINGS index_granularity = 8192
```

Each table has near 20 millions rows, data inserts from Materialized View and Kafka Engine
I'm trying to get count of rows in table `session` which uid not presented in table `slice`
But result count of rows always different, even with `FINAL` keyword use
My query is
```
SELECT COUNT() FROM session AS d1 FINAL WHERE d1.uid NOT IN (SELECT DISTINCT d2.uid FROM slice AS d2 FINAL);
```

Examples of output
```
6446ab0de0c3 :) SELECT COUNT() FROM session AS d1 FINAL WHERE d1.uid NOT IN (SELECT DISTINCT d2.uid FROM slice AS d2 FINAL);

SELECT COUNT()
FROM session AS d1
FINAL
WHERE d1.uid NOT IN (
    SELECT DISTINCT d2.uid
    FROM slice AS d2
    FINAL
)

Query id: dd1d387a-1c95-4f9d-859d-5fd38e9e199a

┌─count()─┐
│    3502 │
└─────────┘

1 rows in set. Elapsed: 9.317 sec. Processed 42.95 million rows, 1.36 GB (4.61 million rows/s., 145.89 MB/s.)

6446ab0de0c3 :) SELECT COUNT() FROM session AS d1 FINAL WHERE d1.uid NOT IN (SELECT DISTINCT d2.uid FROM slice AS d2 FINAL);

SELECT COUNT()
FROM session AS d1
FINAL
WHERE d1.uid NOT IN (
    SELECT DISTINCT d2.uid
    FROM slice AS d2
    FINAL
)

Query id: 5e14eb27-29b7-4f42-8892-8b35b8f9dcfa

┌─count()─┐
│       0 │
└─────────┘

1 rows in set. Elapsed: 9.315 sec. Processed 42.96 million rows, 1.36 GB (4.61 million rows/s., 145.96 MB/s.)

6446ab0de0c3 :) SELECT COUNT() FROM session AS d1 FINAL WHERE d1.uid NOT IN (SELECT DISTINCT d2.uid FROM slice AS d2 FINAL);

SELECT COUNT()
FROM session AS d1
FINAL
WHERE d1.uid NOT IN (
    SELECT DISTINCT d2.uid
    FROM slice AS d2
    FINAL
)

Query id: 43a88150-b3f6-4ad7-b325-fec559d1b407

┌─count()─┐
│    2314 │
└─────────┘

1 rows in set. Elapsed: 9.557 sec. Processed 42.97 million rows, 1.36 GB (4.50 million rows/s., 142.27 MB/s.)
```

**Which ClickHouse server version to use**

There is standalone server with only one shard and replica

```
Connected to ClickHouse server version 22.1.3 revision 54455.
```

**Expected behavior**
Same results of SELECT COUNT queries for each times execution

**Comment by alexeykobzev** [comment]
My apologies
I'm facing problem with setting `do_not_merge_across_partitions_select_final`. It should be set to **1** and problem with zero rows count is gone out
Also I think another mismatch related to Kafka insert in different time by differrent MV, even the same topic used. Correct me if this is not true, but it should be expected behavior

**Comment by den-crane** [comment]
Yes, 

2 problems
1. inserts using Materialized Views expose data for selects in different moments of time.
2. `SELECT COUNT() FROM session in (select` sees different time state for these 2 selects, it even possible with 1 table.

both should be mitigated by transactions.

**Comment by alexeykobzev** [comment]
> both should be mitigated by transactions
AFAIK, currently Clickhouse does not support transactions. What are you meaning?
Do you know about any plans to implementation transactions in Clickhouse?

**Comment by alexeykobzev** [comment]
So I did some research on Github and found that this feature not full implemented yet, but work in progress
https://github.com/ClickHouse/ClickHouse/issues/22086
https://github.com/ClickHouse/ClickHouse/pull/24258
Am I correct?

**Comment by tavplubix** [comment]
Yes, this feature is available since 22.4 as experimental (compatibility may be broken; not recommended for production), development is still in progress.
